### PR TITLE
VS-143: Enable the Analyzer to Display Alerts for Queries that Do NOT work on EF Provider

### PR DESCRIPTION
**Pull request by Ravi-Raghavan**
Enable the Analyzer to Display Alerts for Queries that Do NOT work on EF Provider

### CODE REVIEW

File: tests/MongoDB.Analyzer.Tests.Common.TestCases/EF/NotSupportedEFExpressions.cs [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    [NotSupportedEF("GroupBy Not Supported in EF.", DriverVersions.Linq3OrGreater)]
+    [NotSupportedEF("GroupBy Not Supported in EF.", DriverVersions.Linq3OrGreater)]
+    public void GroupBy()
+    {
+        var dbContextOptions = new DbContextOptionsBuilder<MyDbContext2>();
+        var db = new MyDbContext2(dbContextOptions.Options);
+        var users_query = db.Users.GroupBy(u => u.Address);
+        var customers_query = db.Customers.GroupBy(c => c.LastName);
```
Comment (BorisDog) [inline tests/MongoDB.Analyzer.Tests.Common.TestCases/EF/NotSupportedEFExpressions.cs:66]: Please also add a test where `GroupBy`, `Include` are chained to  a valid simple expression, like `db.Customers.Select.GroupBy`

**Review by BorisDog — CHANGES_REQUESTED** [review]

File: src/MongoDB.Analyzer/Core/Linq/LinqExpressionProcessor.cs [hunk 2]
```diff
@@ -172,4 +173,70 @@ private static bool PreanalyzeLinqExpression(SyntaxNode linqExpressionNode, Sema
...
+    private static bool PreanalyzeEFExpression(SyntaxNode efExpressionNode, SemanticModel semanticModel, List<InvalidExpressionAnalysisNode> invalidEFExpressionNodes, INamedTypeSymbol mongoQueryableNamedType)
+    {
+        var result = true;
+
+        var typeArgument = mongoQueryableNamedType.TypeArguments[0];
+        foreach (var member in typeArgument.GetMembers())
+        {
+            if (member is IPropertySymbol propertySymbol)
+            {
+                //Check TypeArgument for Binary/Byte Array Properties
```
[the remark this answers is not in the document]
Comment (Ravi-Raghavan) [reply src/MongoDB.Analyzer/Core/Linq/LinqExpressionProcessor.cs:186]: Done

File: tests/MongoDB.Analyzer.Tests.Common.TestCases/EF/NotSupportedEFExpressions.cs [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    {
+        var dbContextOptions = new DbContextOptionsBuilder<MyDbContext2>();
+        var db = new MyDbContext2(dbContextOptions.Options);
+
+        _ = db.Users
+                .Join(
+                    db.Customers,
+                    user => user.Age,
+                    customer => customer.CustomerId,
+                    (user, customer) => new
```
[the remark this answers is not in the document]
Comment (Ravi-Raghavan) [reply tests/MongoDB.Analyzer.Tests.Common.TestCases/EF/NotSupportedEFExpressions.cs:88]: Done

File: src/MongoDB.Analyzer/Core/Linq/LinqExpressionProcessor.cs [hunk 4]
```diff
@@ -172,4 +173,70 @@ private static bool PreanalyzeLinqExpression(SyntaxNode linqExpressionNode, Sema
...
+                    invalidEFExpressionNodes.Add(new(
+                                methodInvocation,
+                                EFAnalysisErrorMessages.GroupByMethodNotSupported));
+
+                    result = false;
+                    break;
+                }
+
+                //Select Methods(i.e. Client Side Projections)
+                if (methodSymbol.Name == "Select")
```
[the remark this answers is not in the document]
Comment (Ravi-Raghavan) [reply src/MongoDB.Analyzer/Core/Linq/LinqExpressionProcessor.cs:228]: Done

File: tests/MongoDB.Analyzer.Tests.Common.TestCases/EF/NotSupportedEFExpressions.cs [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+                    {
+                        Age = user.Age,
+                        ID = customer.CustomerId
+                    });
+    }
+}
+
+public class MyDbContext2 : DbContext
```
[the remark this answers is not in the document]
Comment (Ravi-Raghavan) [reply tests/MongoDB.Analyzer.Tests.Common.TestCases/EF/NotSupportedEFExpressions.cs:96]: Done

File: src/MongoDB.Analyzer/Core/Linq/LinqExpressionProcessor.cs [hunk 6]
```diff
@@ -172,4 +173,70 @@ private static bool PreanalyzeLinqExpression(SyntaxNode linqExpressionNode, Sema
...
+        foreach (var member in typeArgument.GetMembers())
+        {
+            if (member is IPropertySymbol propertySymbol)
+            {
+                //Check TypeArgument for Binary/Byte Array Properties
+                if (propertySymbol.Type is IArrayTypeSymbol arrayTypeSymbol &&
+                    arrayTypeSymbol.ElementType.SpecialType == SpecialType.System_Byte)
+                {
+                    invalidEFExpressionNodes.Add(new(
+                                efExpressionNode,
```
[the remark this answers is not in the document]
Comment (Ravi-Raghavan) [reply src/MongoDB.Analyzer/Core/Linq/LinqExpressionProcessor.cs:191]: Done

File: tests/MongoDB.Analyzer.Tests.Common.TestCases/EF/NotSupportedEFExpressions.cs (hunk 1 above)
Comment (Ravi-Raghavan) [reply tests/MongoDB.Analyzer.Tests.Common.TestCases/EF/NotSupportedEFExpressions.cs:66]: Done

File: tests/MongoDB.Analyzer.Tests.Common.TestCases/EF/NotSupportedEFExpressions.cs [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+using System.Linq;
+using Microsoft.EntityFrameworkCore;
+using MongoDB.Analyzer.Tests.Common.DataModel;
+using MongoDB.Driver;
+
+namespace MongoDB.Analyzer.Tests.Common.TestCases.EF;
+
+public sealed class NotSupportedEFExpressions : TestCasesBase
+{
+    [NoDiagnostics]
```
Comment (BorisDog) [inline tests/MongoDB.Analyzer.Tests.Common.TestCases/EF/NotSupportedEFExpressions.cs:24]: Let's remove this.

**Review by BorisDog — CHANGES_REQUESTED** [review]

File: tests/MongoDB.Analyzer.Tests.Common.TestCases/EF/NotSupportedEFExpressions.cs (hunk 7 above)
Comment (Ravi-Raghavan) [reply tests/MongoDB.Analyzer.Tests.Common.TestCases/EF/NotSupportedEFExpressions.cs:24]: Done

File: src/MongoDB.Analyzer/Core/Linq/LinqAnalysisConstants.cs [hunk 8]
```diff
@@ -25,3 +25,9 @@ internal static class LinqAnalysisErrorMessages
 {
     public const string MethodInvocationNotSupported = "Method referencing lambda parameter is not supported LINQ expression.";
 }
+
+internal static class EFAnalysisErrorMessages
+{
+    public const string ByteArraysNotSupported = "Byte Array Property Not Supported.";
+    public const string GroupByMethodNotSupported = "GroupBy Not Supported in EF.";
```
Comment (BorisDog) [inline src/MongoDB.Analyzer/Core/Linq/LinqAnalysisConstants.cs:32]: Not need to capitalize every word.
Consider
'Byte array type is not supported by this version of the EF provider.".
'GroupBy is not supported by this version of the EF provider.".

**Review by BorisDog — CHANGES_REQUESTED** [review]

File: src/MongoDB.Analyzer/Core/Linq/LinqAnalysisConstants.cs (hunk 8 above)
Comment (Ravi-Raghavan) [reply src/MongoDB.Analyzer/Core/Linq/LinqAnalysisConstants.cs:32]: Done

**Review by BorisDog — APPROVED** [review]
LGTM!

**Pull request merged**